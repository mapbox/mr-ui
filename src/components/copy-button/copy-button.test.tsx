import React from 'react';
import CopyButton from './copy-button';
import { act, render, waitFor, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

describe('CopyButton', () => {

  describe('basic', () => {
    const props = {
      text: 'Copied by the default test case'
    };

    test('renders as expected', () => {
      const { baseElement } = render(<CopyButton {...props} />)
      expect(baseElement).toMatchSnapshot();
    });

    test('changes state on click', async () => {
      render(<CopyButton {...props} />)

      await act(async () => {
        await userEvent.click(screen.getByTestId('copy-button'));
      });

      await waitFor(() => {
        expect(screen.getByRole('dialog')).toBeInTheDocument();
      });

      await waitFor(() => {
        expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
      });
    });
  });

  describe('all props', () => {
    const mockOnCopy = jest.fn();
    const props = {
      text: 'more copiable text',
      onCopy: mockOnCopy,
      className: 'px6 py6 btn btn--purple btn--stroke',
      focusTrapPaused: true,
      block: true,
      passthroughProps: {
        'data-testid': 'foo'
      }
    };

    test('renders as expected', () => {
      const { baseElement } = render(<CopyButton {...props} />)
      expect(baseElement).toMatchSnapshot();
    });

    test('changes state on click', async () => {
      render(<CopyButton {...props} />)

      await act(async () => {
        await userEvent.click(screen.getByTestId('copy-button'));
      });

      await waitFor(() => {
        expect(screen.getByRole('dialog')).toBeInTheDocument();
      });

      await waitFor(() => {
        expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
      });
    });

    test('calls onCopy callback', async () => {
      render(<CopyButton {...props} />)

      await act(async () => {
        await userEvent.click(screen.getByTestId('copy-button'));
      });

      expect(mockOnCopy).toHaveBeenCalledTimes(1);
      expect(mockOnCopy).toHaveBeenCalledWith('more copiable text');
    });
  });
});
