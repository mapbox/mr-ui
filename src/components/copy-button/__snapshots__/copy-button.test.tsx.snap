// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CopyButton all props renders as expected 1`] = `
<body>
  <div>
    <div
      aria-label="Copy"
      data-clipboard-text="more copiable text"
      data-testid="copy-button"
      role="button"
    >
      <span
        style="display: contents;"
      >
        <div
          aria-controls="radix-4"
          aria-expanded="false"
          aria-haspopup="dialog"
          data-state="closed"
          type="button"
        >
          <button
            class="px6 py6 btn btn--purple btn--stroke block"
            data-state="closed"
            data-testid="foo"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="events-none icon"
              data-testid="icon-clipboard"
              focusable="false"
              style="width: 18px; height: 18px;"
            >
              <use
                xlink:href="#icon-clipboard"
                xmlns:xlink="http://www.w3.org/1999/xlink"
              />
            </svg>
            <span
              style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
            >
              clipboard
            </span>
          </button>
        </div>
      </span>
    </div>
  </div>
</body>
`;

exports[`CopyButton basic renders as expected 1`] = `
<body>
  <div>
    <div
      aria-label="Copy"
      data-clipboard-text="Copied by the default test case"
      data-testid="copy-button"
      role="button"
    >
      <span
        style="display: contents;"
      >
        <div
          aria-controls="radix-0"
          aria-expanded="false"
          aria-haspopup="dialog"
          data-state="closed"
          type="button"
        >
          <button
            class="btn btn--xs py3 px3 round"
            data-state="closed"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="events-none icon"
              data-testid="icon-clipboard"
              focusable="false"
              style="width: 18px; height: 18px;"
            >
              <use
                xlink:href="#icon-clipboard"
                xmlns:xlink="http://www.w3.org/1999/xlink"
              />
            </svg>
            <span
              style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
            >
              clipboard
            </span>
          </button>
        </div>
      </span>
    </div>
  </div>
</body>
`;
